Feature: catch method

patinthehat opened the pull request:
### Summary

This PR adds a new method: `catch`, which allows chainable, flexible handling of exceptions when passing a callable to `ray()`.

`catch` accepts a variety of arguments for handling exceptions:
- No arguments - all exceptions are reported using the `exception()` function
- A callable accepting a single untyped `$exception` parameter - execute the callable to handle any Exception as needed
- A callable accepting a single type-hinted `$exception` parameter - execute the callable to handle only the specified Exception class as needed
- A string containing the classname of an Exception to be handled (sent to `exception()`)
- An array containing any combination of the previous items

This PR also includes unit tests and updated documentation.

### Rationale
I feel that adding a method for handling exceptions within the `ray()` context will enhance the overall developer experience by allowing developers to write fewer lines of code that is easy to reason about.  Using a simple `catch` method that intuitively works and ensures that functionality can be  easily understood by developers adds value to the Ray application and packages.

### Examples

```php
ray($callable)->catch();
// execution will continue. 
```

You can also pass a callable to `catch` to customize the handling of an Exception.  If you typehint the `$exception` variable, only Exceptions of that type will be handled.  PHP 8 union types are supported.

```php
ray($callable)->catch(function(MyException $exception) {
   // do something with $exception if it is of the MyException type 
});

ray($callable)->catch(function($exception) {
   // handle any exception type
});

ray($callable)->catch(MyException::class);

ray($callable)->catch([
    ExceptionOne::class,
    ExceptionTwo::class,
]);
```

### Demo code
```php
class ExceptionA extends \Exception { }
class ExceptionB extends \Exception { }
class ExceptionC extends \Exception { }

function myfunc($value) {
    if ($value > 8) {
        throw new ExceptionC("Exception C {$value}");        
    }
    
    if ($value > 6) {
        throw new ExceptionB("Exception B {$value}");
    }

    if ($value > 3) {
        throw new ExceptionA("Exception A {$value}");
    }
        
    return "hello world {$value}";
}


foreach(range(1, 9) as $num) {
    $callable = function() use ($num) { return myfunc($num); };

    ray($callable)->catch([
        fn(ExceptionB $e) => ray()->send($e)->green(),
        fn(ExceptionA $e) => ray()->text($e->getMessage())->blue(),
    ])->small();
}
```
![image](https://user-images.githubusercontent.com/5508707/129813468-ca5bbf1d-faca-4259-8604-6c513621c3a7.png)

Following this output in Ray, the following output is displayed in the terminal:

```
PHP Fatal error:  Uncaught ExceptionC: Exception C 9 in /.../test.php:20
Stack trace:
...
```

freekmurze commented on docs/usage/framework-agnostic-php-project.md:
```diff
@@ -607,6 +607,46 @@ try {
+}
+
+foreach(range(1, 10) as $num) {
+    ray()->try(function($ray) use ($num) {
```
I'm thinking we should use the most simple example here. I feel that using the passed `$ray` instance is rather exotic. Let's just use the `ray()` function in the callbacks to send stuff to Ray.

patinthehat commented:
ping @freekmurze

freekmurze commented:
I'll go over this one next week 👍

freekmurze commented:
Thank you for this. 

I think we could simplify this. If developers need fine grained control over exceptions, by just using regular `try/catch`.


We could however add a handy catch method. If there's an exception in the given callable, then `catch` will catch that exception and the execution will continue.

```php
ray($callablle)->catch();
// execution will continue.
```

You could also pass a `callable to catch, which receives the Exception, so handling can be customised.

```php
ray($callable)->catch(function(Exception $exception) {
   // do something with $exception
});

Bonus point can be earned with these:

- if a string is passed to `catch` then it will only swallow the exceptions of that type.

```
ray($callable()->catch(MyCustomException::class);
// execution will continue if there was no exception or the exception is of type `MyCustomException`
```


- if a callable is passed with a type hint, then it will only shallow exceptions of that type.

```php
ray($callable()->catch(function(MyCustomException|MyOtherException) {
   // do something
});
// execution will continue if there was no exception or the exception is of type `MyCustomException` or `MyOtherException`
```

patinthehat commented:
> ```
> I think we could simplify this.
> ```

I agree.  I'll refactor the implementation with those features. :rocket:

patinthehat commented:
@freekmurze I've implemented the simplified functionality we discussed above, removed the `try` method, and updated the PR description/docs/tests accordingly.  Thoughts?

freekmurze commented on docs/usage/framework-agnostic-php-project.md:
````diff
@@ -607,6 +607,59 @@ try {
+]);
+```
+
+If you would like to immediately throw any unhandled exceptions from the callable after calling `ray`, chain the `throwExceptions` function onto the `ray` call.  If `throwExceptions` is not chained, it will be called when PHP finishes executing the script or application.
````
Could you add a code example of this as well?

freekmurze commented on src/Ray.php:
```diff
@@ -568,6 +597,32 @@ public function send(...$arguments): self
+                // use a specific class we can filter out instead of null so that null
+                // payloads can still be sent.
+                return $result instanceof Ray ? IgnoredValue::make() : $result;
+            } catch (\Exception $e) {
```
Import this

freekmurze replied:
Do not use abbreviations: rename `$e` to `$exception`

freekmurze commented on src/Ray.php:
```diff
@@ -568,6 +597,32 @@ public function send(...$arguments): self
+            }
+        }, $arguments);
+
+        $arguments = array_filter($arguments, function ($arg) {
```
Rename `arg` to `argument`

freekmurze commented on src/Support/ExceptionHandler.php:
```diff
@@ -0,0 +1,124 @@
+                    if (is_a($caughtException, $expectedClass, true)) {
+                        $isExpected = true;
+
+                        break 2;
```
Could we rewrite this to avoid having that `break 2;` Maybe a function that loops through these things and returns a boolean that determines if we should continue

freekmurze commented on src/Support/ExceptionHandler.php:
```diff
@@ -0,0 +1,124 @@
+        return $ray->exception($exception);
+    }
+
+    protected function processCallback(Ray $ray, $callback, $rethrow = true): Ray
```
This is a rather long function. Could we break this up in smaller functions with a good descriptive name?

freekmurze commented on src/Support/ExceptionHandler.php:
```diff
@@ -0,0 +1,124 @@
+
+        $result = is_a($paramType, '\\ReflectionUnionType') ? $paramType->getTypes() : [$paramType->getName()];
+
+        return array_map(function ($type) {
```
Use a short function here, then this can be a one liner.

patinthehat replied:
@freekmurze Since arrow functions were added in 7.4 and `spatie/ray` supports 7.3, we have to use a regular function, no?

freekmurze commented on tests/RayTest.php:
```diff
@@ -1099,6 +1099,171 @@ public function it_sends_a_payload_once_while_allowing_calls_to_limit()
+    public function it_handles_exceptions_using_catch_with_an_array_of_callbacks_with_typed_parameters()
+    {
+        $this->getNewRay()->send(function () {
+            throw new \InvalidArgumentException('test');
```
Import everything (also in other files in this PR.

freekmurze commented:
Thanks!

patinthehat commented:
@freekmurze I didn't get a chance to finish all requested changes, I'll submit a secondary PR with them shortly.

freekmurze commented:
Oh, I didn't notice that, I was merging optimistically 🙂
